File: aluno-webrtc-ios/aluno-webrtc-ios/services/WebRTCService.swift
//
//  WebRTCService.swift
//  aluno-webrtc-ios
//
//  Created by Isequiel Henrique do Nascimento on 20/08/25.
//
// WebRTCService.swift

import Foundation
import WebRTC
import SocketIO

@MainActor
class WebRTCService: NSObject, ObservableObject {
    
    private var socket: SocketIOClient?
    private var manager: SocketManager?
    private var peerConnection: RTCPeerConnection?
    private var teacherSocketId: String?
    
    // Mude de RTCMTLVideoView? para RTCVideoTrack?
    @Published var remoteVideoTrack: RTCVideoTrack?

    private let factory: RTCPeerConnectionFactory = {
        RTCInitializeSSL()
        let videoEncoderFactory = RTCDefaultVideoEncoderFactory()
        let videoDecoderFactory = RTCDefaultVideoDecoderFactory()
        return RTCPeerConnectionFactory(encoderFactory: videoEncoderFactory, decoderFactory: videoDecoderFactory)
    }()


    static func checkRoomExists(serverUrl: String, roomCode: String) async throws -> Bool {
         guard let url = URL(string: "\(serverUrl)/room/\(roomCode)") else {
             throw URLError(.badURL)
         }
         
         let (data, response) = try await URLSession.shared.data(from: url)
         
         guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
             throw URLError(.badServerResponse )
         }
         
         guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
               let exists = json["exists"] as? Bool else {
             throw URLError(.cannotParseResponse)
         }
         
         return exists
     }

    func joinRoom(serverUrl: String, roomCode: String) {
        manager = SocketManager(socketURL: URL(string: serverUrl)!, config: [.log(true), .compress])
        socket = manager?.defaultSocket

        setupSocketListeners(roomCode: roomCode)
        socket?.connect()
    }
    
    private func setupSocketListeners(roomCode: String) {
        socket?.on(clientEvent: .connect) { [weak self] _, _ in
           // print("Conectado no servidor socket")
            self?.socket?.emit("student-join", ["roomCode": roomCode])
        }

        // Recebendo a offer do professor
        socket?.on("offer") { [weak self] data, _ in
            guard let self = self,
                  let dict = data.first as? [String: Any],
                  let offerDict = dict["offer"] as? [String: Any],
                  let sdp = offerDict["sdp"] as? String,
                  let senderId = dict["senderId"] as? String else { return }

            //  print("Oferta recebida do professor: \(senderId)")

            // Salva o ID do professor para enviar a resposta e os candidatos ICE de volta
            self.teacherSocketId = senderId
            
            Task {
                await self.handleOffer(sdp: sdp)
            }
        }

        // Recebendo candidatos ICE do professor
        socket?.on("ice-candidate") { [weak self] data, _ in
            guard let self = self,
                  let dict = data.first as? [String: Any],
                  let candidateDict = dict["candidate"] as? [String: Any],
                  let candidateSdp = candidateDict["candidate"] as? String,
                  let sdpMid = candidateDict["sdpMid"] as? String,
                  let sdpMLineIndex = candidateDict["sdpMLineIndex"] as? Int32 else {
               // print("Erro ao decodificar candidato ICE recebido")
                return
            }
            
            print("📬 Candidato ICE recebido")
            let iceCandidate = RTCIceCandidate(sdp: candidateSdp, sdpMLineIndex: sdpMLineIndex, sdpMid: sdpMid)
            
            Task {
                do {
                    try await self.peerConnection?.add(iceCandidate)
                } catch {
                    print("❌ Erro ao adicionar candidato ICE recebido: \(error)")
                }
            }
        }
        
    }
    
    private func createPeerConnection() async throws {
        let config = RTCConfiguration()
        config.iceServers = [RTCIceServer(urlStrings: ["stun:stun.l.google.com:19302"])]
        
        let constraints = RTCMediaConstraints(mandatoryConstraints: nil, optionalConstraints: nil)
        
        guard let pc = factory.peerConnection(with: config, constraints: constraints, delegate: self) else {
            throw NSError(domain: "WebRTCService", code: -1, userInfo: [NSLocalizedDescriptionKey: "Falha ao criar PeerConnection"])
        }
        self.peerConnection = pc
    }

    private func handleOffer(sdp: String) async {
        do {

            // 1. Criar a PeerConnection se ainda não existir
            if peerConnection == nil {
                try await createPeerConnection()
            }
            
            // 2. Definir a oferta recebida como descrição remota
            let remoteDescription = RTCSessionDescription(type: .offer, sdp: sdp)
            try await peerConnection?.setRemoteDescription(remoteDescription)
        //    print(" Descrição remota (oferta) definida.")

            // 3. Criar uma resposta (answer)
            let constraints = RTCMediaConstraints(mandatoryConstraints: [
                "OfferToReceiveAudio": "false",
                "OfferToReceiveVideo": "true"
            ], optionalConstraints: nil)
            let answer = try await peerConnection?.answer(for: constraints)
            
            // 4. Definir a resposta como descrição local
            guard let answer = answer else { return }
            try await peerConnection?.setLocalDescription(answer)
        //    print(" Descrição local (resposta) definida.")

            // 5. Enviar a resposta de volta para o professor
            guard let teacherId = self.teacherSocketId else { return }
            let answerSdp = ["sdp": answer.sdp, "type": "answer"]
            socket?.emit("answer", ["answer": answerSdp, "targetId": teacherId])
        //   print(" Resposta enviada para o professor \(teacherId)")

        } catch {
        //    print(" Erro no processamento da offer: \(error)")
        }
    }
}

extension WebRTCService: RTCPeerConnectionDelegate {

    // cahamada quando um candidato ICE local é gerado.
    func peerConnection(_ peerConnection: RTCPeerConnection, didGenerate candidate: RTCIceCandidate) {
        guard let teacherId = self.teacherSocketId else {
    //        print(" Candidato ICE gerado, mas o ID do professor não está disponível.")
            return
        }
        
        let candidateData: [String: Any] = [
            "candidate": candidate.sdp,
            "sdpMid": candidate.sdpMid ?? "",
            "sdpMLineIndex": candidate.sdpMLineIndex
        ]
        
        // Envia o candidato local para o servidor, direcionado ao professor
        socket?.emit("ice-candidate", [
            "candidate": candidateData,
            "targetId": teacherId
        ])
    
    }

    func peerConnection(_ peerConnection: RTCPeerConnection, didAdd stream: RTCMediaStream) {
        print("📺 Stream remoto recebido com \(stream.videoTracks.count) vídeos e \(stream.audioTracks.count) áudios")
        if let track = stream.videoTracks.first {
            // Publica o RTCVideoTrack para a View
            DispatchQueue.main.async {
                self.remoteVideoTrack = track
            }
        }
    }
    
    func peerConnection(_ peerConnection: RTCPeerConnection, didChange newState: RTCIceConnectionState) {
    }

    // Funções de delegate
    func peerConnection(_ peerConnection: RTCPeerConnection, didRemove stream: RTCMediaStream) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didChange stateChanged: RTCSignalingState) {}
    func peerConnectionShouldNegotiate(_ peerConnection: RTCPeerConnection) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didChange newState: RTCIceGatheringState) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didRemove candidates: [RTCIceCandidate]) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didOpen dataChannel: RTCDataChannel) {}
}

File: student_flutter_app/lib/services/webrtc_service.dart
import 'dart:convert';
import 'package:flutter_webrtc/flutter_webrtc.dart';
import 'package:socket_io_client/socket_io_client.dart' as IO;
import 'package:http/http.dart' as http;

class WebRTCService {
  IO.Socket? _socket;
  RTCPeerConnection? _peerConnection;
  
  // Callbacks
  Function(RTCPeerConnectionState)? onConnectionStateChanged;
  Function(MediaStream)? onRemoteStream;
  Function()? onStreamEnded;
  Function()? onTeacherDisconnected;
  Function(String)? onError;

  // Configuração ICE servers
final Map<String, dynamic> _iceServers = {
  'iceServers': [
    {'urls': 'stun:stun.l.google.com:19302'},
    {
      'urls': 'turn:meu-turn-server.com:3478',
      'username': 'usuario',
      'credential': 'senha'
    }
  ]
};


  Future<bool> checkRoomExists(String serverUrl, String roomCode) async {
    try {
      final response = await http.get(
        Uri.parse('$serverUrl/room/$roomCode'),
        headers: {'Content-Type': 'application/json'},
      ).timeout(const Duration(seconds: 10));

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return data['exists'] == true;
      }
      return false;
    } catch (e) {
      throw Exception('Erro ao verificar sala: $e');
    }
  }

  Future<void> joinRoom(String serverUrl, String roomCode) async {
    try {
      // Conectar ao servidor WebSocket
      _socket = IO.io(serverUrl, <String, dynamic>{
        'transports': ['websocket'],
        'autoConnect': false,
      });

      _setupSocketListeners(roomCode);
      _socket!.connect();

      // Aguardar conexão
      await Future.delayed(const Duration(seconds: 2));
      
      if (!_socket!.connected) {
        throw Exception('Não foi possível conectar ao servidor');
      }

      // Entrar na sala 
      _socket!.emit('student-join', {'roomCode': roomCode});

    } catch (e) {
      onError?.call('Erro ao entrar na sala: $e');
    }
  }

  void _setupSocketListeners(String roomCode) {
    _socket!.on('connect', (_) {
      print('Conectado ao servidor WebSocket');
    });

    _socket!.on('disconnect', (_) {
      print('Desconectado do servidor WebSocket');
      onError?.call('Conexão perdida com o servidor');
    });

    _socket!.on('joined-room', (data) {
      print('Entrou na sala: ${data['roomCode']}');
    });

    _socket!.on('stream-started', (_) {
      print('Transmissão iniciada pelo professor');
      // A oferta será recebida em seguida
    });

    _socket!.on('stream-stopped', (_) {
      print('Transmissão encerrada pelo professor');
      onStreamEnded?.call();
      _closePeerConnection();
    });

    _socket!.on('teacher-disconnected', (_) {
      print('Professor desconectado');
      onTeacherDisconnected?.call();
      _closePeerConnection();
    });

    _socket!.on('offer', (data) async {
      print('Oferta recebida do professor');
      await _handleOffer(data['offer'], data['senderId']);
    });

    _socket!.on('ice-candidate', (data) async {
      print('Candidato ICE recebido');
      await _handleIceCandidate(data['candidate']);
    });

    _socket!.on('error', (data) {
      print('Erro do servidor: ${data['message']}');
      onError?.call(data['message']);
    });
  }

  Future<void> _handleOffer(Map<String, dynamic> offer, String senderId) async {
    try {
      // Criar peer connection se não existir
      if (_peerConnection == null) {
        await _createPeerConnection();
      }

      // Definir descrição remota
      await _peerConnection!.setRemoteDescription(
        RTCSessionDescription(offer['sdp'], offer['type']),
      );

      // Criar resposta
      RTCSessionDescription answer = await _peerConnection!.createAnswer();
      await _peerConnection!.setLocalDescription(answer);

      // Enviar resposta
      _socket!.emit('answer', {
        'answer': answer.toMap(),
        'targetId': senderId,
      });

      print('Resposta enviada para o professor');

    } catch (e) {
      print('Erro ao processar oferta: $e');
      onError?.call('Erro ao processar oferta do professor');
    }
  }

  Future<void> _handleIceCandidate(Map<String, dynamic>? candidate) async {
    if (_peerConnection != null && candidate != null) {
      try {
        await _peerConnection!.addCandidate(
          RTCIceCandidate(
            candidate['candidate'],
            candidate['sdpMid'],
            candidate['sdpMLineIndex'],
          ),
        );
      } catch (e) {
        print('Erro ao adicionar candidato ICE: $e');
      }
    }
  }

  Future<void> _createPeerConnection() async {
    try {
      _peerConnection = await createPeerConnection(_iceServers);

      // Configurar eventos
      _peerConnection!.onConnectionState = (state) {
        print('Estado da conexão: $state');
        onConnectionStateChanged?.call(state);
      };

      _peerConnection!.onAddStream = (stream) {
        print('Stream remoto recebido');
        onRemoteStream?.call(stream);
      };

      _peerConnection!.onIceCandidate = (candidate) {
        print('Candidato ICE local gerado');
        _socket!.emit('ice-candidate', {
          'candidate': candidate.toMap(),
          'targetId': null, // Enviar para o professor
        });
      };

      _peerConnection!.onIceConnectionState = (state) {
        print('Estado da conexão ICE: $state');
        if (state == RTCIceConnectionState.RTCIceConnectionStateFailed ||
            state == RTCIceConnectionState.RTCIceConnectionStateDisconnected) {
          onError?.call('Conexão perdida com o professor');
        }
      };

    } catch (e) {
      print('Erro ao criar peer connection: $e');
      onError?.call('Erro ao estabelecer conexão WebRTC');
    }
  }

  void _closePeerConnection() {
    _peerConnection?.close();
    _peerConnection = null;
  }

  void dispose() {
    _closePeerConnection();
    _socket?.disconnect();
    _socket?.dispose();
  }
}

